json/token.go

```diff
@@ -76,6 +76,9 @@ type Tokenizer struct {
 
 	// Stack used to track entering and leaving arrays, objects, and keys.
 	stack *stack
+
+	// Parsing flags.
+	flags ParseFlags
```

Comment:
Do you think we could have an API to expose this information to the program? That way when decoding the `RawValue` into a string, the program can avoid calling `ascii.ValidPrint` + `strings.IndexByte` again. Maybe something like this:

```go
// Unquote unquotes the string value that the tokenizer is positioned on.
// As an optimization, the tokenizer may avoid allocating a new string if
// the current value did not need to be unescaped.
func (t *Tokenizer) Unquote() []byte {
  ...
}
```

Reply:
Is it possible to tag the `RawValue`? That way you can pass on info from the `parseString` call and just slice or copy the buffer when calling `RawValue.Unquote()`.

This change gives me:

```
TokenizerUnquote-4   139MB/s ± 2%   337MB/s ± 0%  +141.82%  (p=0.008 n=5+5)
```

```diff
diff --git i/json/token.go w/json/token.go
index 5f10c6f..a95854c 100644
--- i/json/token.go
+++ w/json/token.go
@@ -100,7 +100,7 @@ func (t *Tokenizer) Reset(b []byte) {
 	//
 	// However, it does not compile down to an invocation of duff-copy.
 	t.Delim = 0
-	t.Value = nil
+	t.Value = RawValue{}
 	t.Err = nil
 	t.Depth = 0
 	t.Index = 0
@@ -144,27 +144,33 @@ skipLoop:
 		return false
 	}
 
+	t.Value.flags = 0
+
 	switch t.json[0] {
 	case '"':
 		t.Delim = 0
-		t.Value, t.json, t.Err = t.decoder.parseString(t.json)
+		var escaped bool
+		t.Value.B, t.json, escaped, t.Err = t.decoder.parseStringFast(t.json)
+		if !escaped {
+			t.Value.flags |= unquoted
+		}
 	case 'n':
 		t.Delim = 0
-		t.Value, t.json, t.Err = t.decoder.parseNull(t.json)
+		t.Value.B, t.json, t.Err = t.decoder.parseNull(t.json)
 	case 't':
 		t.Delim = 0
-		t.Value, t.json, t.Err = t.decoder.parseTrue(t.json)
+		t.Value.B, t.json, t.Err = t.decoder.parseTrue(t.json)
 	case 'f':
 		t.Delim = 0
-		t.Value, t.json, t.Err = t.decoder.parseFalse(t.json)
+		t.Value.B, t.json, t.Err = t.decoder.parseFalse(t.json)
 	case '-', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9':
 		t.Delim = 0
-		t.Value, t.json, t.Err = t.decoder.parseNumber(t.json)
+		t.Value.B, t.json, t.Err = t.decoder.parseNumber(t.json)
 	case '{', '}', '[', ']', ':', ',':
-		t.Delim, t.Value, t.json = Delim(t.json[0]), t.json[:1], t.json[1:]
+		t.Delim, t.Value.B, t.json = Delim(t.json[0]), t.json[:1], t.json[1:]
 	default:
 		t.Delim = 0
-		t.Value, t.json, t.Err = t.json[:1], t.json[1:], syntaxError(t.json, "expected token but found '%c'", t.json[0])
+		t.Value.B, t.json, t.Err = t.json[:1], t.json[1:], syntaxError(t.json, "expected token but found '%c'", t.json[0])
 	}
 
 	t.Depth = t.depth()
@@ -203,7 +209,7 @@ skipLoop:
 		}
 	}
 
-	return (t.Delim != 0 || len(t.Value) != 0) && t.Err == nil
+	return (t.Delim != 0 || len(t.Value.B) != 0) && t.Err == nil
 }
 
 func (t *Tokenizer) depth() int {
@@ -236,24 +242,29 @@ func (t *Tokenizer) pop(expect scope) error {
 
 // RawValue represents a raw json value, it is intended to carry null, true,
 // false, number, and string values only.
-type RawValue []byte
+type RawValue struct {
+	B     []byte
+	flags int
+}
+
+const unquoted = 1 << iota
 
 // String returns true if v contains a string value.
-func (v RawValue) String() bool { return len(v) != 0 && v[0] == '"' }
+func (v RawValue) String() bool { return len(v.B) != 0 && v.B[0] == '"' }
 
 // Null returns true if v contains a null value.
-func (v RawValue) Null() bool { return len(v) != 0 && v[0] == 'n' }
+func (v RawValue) Null() bool { return len(v.B) != 0 && v.B[0] == 'n' }
 
 // True returns true if v contains a true value.
-func (v RawValue) True() bool { return len(v) != 0 && v[0] == 't' }
+func (v RawValue) True() bool { return len(v.B) != 0 && v.B[0] == 't' }
 
 // False returns true if v contains a false value.
-func (v RawValue) False() bool { return len(v) != 0 && v[0] == 'f' }
+func (v RawValue) False() bool { return len(v.B) != 0 && v.B[0] == 'f' }
 
 // Number returns true if v contains a number value.
 func (v RawValue) Number() bool {
-	if len(v) != 0 {
-		switch v[0] {
+	if len(v.B) != 0 {
+		switch v.B[0] {
 		case '-', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9':
 			return true
 		}
@@ -263,8 +274,12 @@ func (v RawValue) Number() bool {
 
 // AppendUnquote writes the unquoted version of the string value in v into b.
 func (v RawValue) AppendUnquote(b []byte) []byte {
+	if (v.flags & unquoted) != 0 {
+		return v.B[1:len(v.B)-1]
+	}
+
 	d := decoder{}
-	s, r, new, err := d.parseStringUnquote(v, b)
+	s, r, new, err := d.parseStringUnquote(v.B, b)
 	if err != nil {
 		panic(err)
 	}
diff --git i/json/token_test.go w/json/token_test.go
index 2805de3..7bc6f0e 100644
--- i/json/token_test.go
+++ w/json/token_test.go
@@ -17,7 +17,7 @@ type token struct {
 func delim(s string, depth, index int) token {
 	return token{
 		delim: Delim(s[0]),
-		value: RawValue(s),
+		value: RawValue{B: []byte(s)},
 		depth: depth,
 		index: index,
 	}
@@ -25,7 +25,7 @@ func delim(s string, depth, index int) token {
 
 func key(v string, depth, index int) token {
 	return token{
-		value: RawValue(v),
+		value: RawValue{B: []byte(v)},
 		depth: depth,
 		index: index,
 		isKey: true,
@@ -34,7 +34,7 @@ func key(v string, depth, index int) token {
 
 func value(v string, depth, index int) token {
 	return token{
-		value: RawValue(v),
+		value: RawValue{B: []byte(v)},
 		depth: depth,
 		index: index,
 	}
@@ -321,3 +321,53 @@ func BenchmarkTokenizer(b *testing.B) {
 		})
 	}
 }
+
+func BenchmarkTokenizerUnquote(b *testing.B) {
+	json := []byte(`{
+		"timestamp": "2019-01-09T18:59:57.456Z",
+		"channel": "server",
+		"type": "track",
+		"event": "Test",
+		"userId": "test-user-whatever",
+		"messageId": "test-message-whatever",
+		"integrations": {
+			"whatever": {
+				"debugMode": false
+			},
+			"myIntegration": {
+				"debugMode": true
+			}
+		},
+		"properties": {
+			"trait1": 1,
+			"trait2": "test",
+			"trait3": true
+		},
+		"settings": {
+			"apiKey": "1234567890",
+			"debugMode": false,
+			"directChannels": [
+				"server",
+				"client"
+			],
+			"endpoint": "https://somewhere.com/v1/integrations/segment"
+		}
+	}`)
+
+	t := NewTokenizer(nil)
+
+	b.SetBytes(int64(len(json)))
+	b.ResetTimer()
+
+	for i := 0; i < b.N; i++ {
+		t.Reset(json)
+		for t.Next() {
+			if t.Value.String() {
+				t.Value.Unquote()
+			}
+		}
+		if t.Err != nil {
+			b.Error(t.Err)
+		}
+	}
+}
```

Reply:
Yeah we could do it this way, but that change would likely break compilation of programs using json.RawValue.

Reply:
It's ok to leave this one change for a follow up as well. It may call for a more coherent approach of how to help programs that use the tokenizer achieve greater efficiency.